python/lsst/ts/salobj/sal_info.py

```diff
@@ -321,31 +354,35 @@ def add_reader(self, topic):
                     self.log.info(f"Read historical data in {dt:0.2f} sec")
                 else:
                     self.log.warning(f"Could not read historical data in {dt:0.2f} sec")
-                for read_cond, topic in list(self._readers.items()):
+
+                # read historical (late-joiner) data
+                for read_cond, reader in list(self._readers.items()):
                     if not self.isopen:  # shutting down
                         return
-                    if not topic.isopen or topic.max_history == 0:
-                        # reader closed or gets no history
+                    if reader.volatile or not reader.isopen or not read_cond.triggered():
+                        # reader gets no historical data, is closed,
+                        # or has no data to be read
                         continue
                     try:
-                        data_list = topic._reader.take_cond(read_cond, DDS_READ_QUEUE_LEN)
+                        data_list = reader._reader.take_cond(read_cond, DDS_READ_QUEUE_LEN)
                     except dds.DDSException as e:
-                        self.log.warning(f"dds error while reading late joiner data for {topic}; "
+                        self.log.warning(f"dds error while reading late joiner data for {reader}; "
                                          f"trying again: {e}")
                         time.sleep(0.001)
-                        data_list = topic._reader.take_cond(read_cond, DDS_READ_QUEUE_LEN)
-                    self.log.debug(f"Read {len(data_list)} history items for {topic}")
+                        try:
+                            data_list = reader._reader.take_cond(read_cond, DDS_READ_QUEUE_LEN)
+                        except dds.DDSException as e:
+                            raise RuntimeError(f"dds error while reading late joiner data for {reader}; "
+                                               "giving up") from e
+                    self.log.debug(f"Read {len(data_list)} history items for {reader}")
                     sd_list = [self._sample_to_data(sd, si) for sd, si in data_list if si.valid_data]
-                    # TODO DM-20313: enable this code
-                    # which was commented out to work around DM-20312:
-                    # if len(sd_list) < len(data_list):
-                    #     ninvalid = len(data_list) - len(sd_list)
-                    #     self.log.warning(f"Bug: read {ninvalid} late joiner "
-                    #                      f"items for {topic}")
-                    if topic.max_history > 0:
-                        sd_list = sd_list[-topic.max_history:]
+                    if len(sd_list) < len(data_list):
+                        ninvalid = len(data_list) - len(sd_list)
+                        self.log.warning(f"Bug: read {ninvalid} late joiner items for {reader}")
```

Comment:
It would be a bug in this code or in OpenSplice dds. I have never seen the message. If I did see it I'd probably have to open an OpenSplice ticket.

Reply:
That sounds like an `error` then. I feel like `warning` can be sometimes lost in the noise. Please paraphrase your above comment into the message for future debuggers.

Reply:
I strongly prefer a warning because the code has handled the condition; it just didn't expect to need to do so. I'll reword the messages to make this clearer in both places. Here is my proposed rewording of the first message:
```
self.log.warning(f"Read {ninvalid} invalid late-joiner items from {reader}. "
                 "The invalid items were safely skipped, but please examine "
                 "the code in SalInfo.start to see if it needs an update "
                 "for changes to OpenSplice dds.")
```

By the way, if this condition never occurs then I can ditch the test entirely. I am being very careful because DDS reads *can* return data marked invalid. It's a feature, but not one that this code needs and I try very hard to avoid it in these reads.

Reply:
The only thing I object to here is "If this condition never occurs". That's... not something one can test for.

Reply:
I think after a month or year of operation one can mine the EFD looking for these messages. If not found then one can update the code.

But I can also imagine leaving the code in forever, in case some newer version of OpenSplice has a regression or a subtle change that results in reading invalid samples.